crates/bevy_math/src/cubic_splines.rs

```diff
@@ -279,13 +290,279 @@ impl<P: Point> CubicGenerator<P> for CubicBSpline<P> {
+    /// start and end points.
+    ///
+    /// The knot vector will have a knot with multiplicity of 4 at the end and start and elements
+    /// in the middle will have a difference of 1. "Multiplicity" means taht there are N
+    /// consecutive elements that have the same value.
+    ///
+    /// Will return `None` if there are less than 4 control points
+    pub fn open_uniform_knot_vector(control_points: usize) -> Option<Vec<f32>> {
+        if control_points < 4 {
+            return None;
+        }
+        let last_knots_value = control_points - 3;
+        Some(
+            std::iter::repeat(0.0)
+                .take(4)
+                .chain((1..last_knots_value).map(|v| v as f32))
+                .chain(std::iter::repeat(last_knots_value as f32).take(4))
+                .collect(),
+        )
+    }
+
+    #[inline(always)]
+    const fn knot_vector_length(control_points_len: usize) -> usize {
+        control_points_len + 4
+    }
+
+    /// Based on <https://xiaoxingchen.github.io/2020/03/02/bspline_in_so3/general_matrix_representation_for_bsplines.pdf>
+    fn generate_matrix(knot_vector_segment: &[f32; 8]) -> [[f32; 4]; 4] {
+        let t = knot_vector_segment;
+        let i = 3;
+        let m00 = (t[i + 1] - t[i]).powi(2) / ((t[i + 1] - t[i - 1]) * (t[i + 1] - t[i - 2]));
+        let m02 = (t[i] - t[i - 1]).powi(2) / ((t[i + 2] - t[i - 1]) * (t[i + 1] - t[i - 1]));
+        let m12 = (3.0 * (t[i + 1] - t[i]) * (t[i] - t[i - 1]))
+            / ((t[i + 2] - t[i - 1]) * (t[i + 1] - t[i - 1]));
+        let m22 = 3.0 * (t[i + 1] - t[i]).powi(2) / ((t[i + 2] - t[i - 1]) * (t[i + 1] - t[i - 1]));
+        let m33 = (t[i + 1] - t[i]).powi(2) / ((t[i + 3] - t[i]) * (t[i + 2] - t[i]));
+        let m32 = -m22 / 3.0
+            - m33
+            - (t[i + 1] - t[i]).powi(2) / ((t[i + 2] - t[i]) * (t[i + 2] - t[i - 1]));
+        [
+            [m00, 1.0 - m00 - m02, m02, 0.0],
+            [-3.0 * m00, 3.0 * m00 - m12, m12, 0.0],
+            [3.0 * m00, -3.0 * m00 - m22, m22, 0.0],
+            [-m00, m00 - m32 - m33, m32, m33],
+        ]
+    }
+
+    /// Normalizes weights vector using L0 norm.
+    /// The resulting weight vector's values will add up to be equal the amoutn of values in teh
+    /// weights vector
+    fn normalize_weights(weights: &mut [f32]) {
+        let g = weights.len() as f32;
+        let weights_sum = weights.iter().fold(0.0, |sum, w| sum + w);
+        let mul = g / weights_sum;
+        weights.iter_mut().for_each(|w| *w *= mul);
+    }
+}
+impl<P: Point> CubicGenerator<P> for CubicNurbs<P> {
+    #[inline]
+    fn to_curve(&self) -> CubicCurve<P> {
+        let segments = self
+            .control_points
+            .windows(4)
+            .enumerate()
+            .map(|(i, points)| {
+                // Unwrap will not fail because the slice is of length 8
+                let knot_vector_segment = (&self.knot_vector[i..i + 8]).try_into().unwrap();
```

Comment:
It might make sense to use `.zip(knot_vector.windows(8))` here instead of this index pattern.

Also `.map(|win| [0,1,2,3,4,5,6,7].map(|i| win[i]))` helps to get rid of the `try_into().unwrap()`

Reply:
Using `try_into.expect("...")` seems to be more explicit for converting a slice that `windows` provides to a statiscally-typed array. The window length is set as a constant number, so we know the size ahead of time.

Reply:
That's also fair! Probably just boils down to taste. 👍